book/multi-arch-controller.md

```diff
@@ -0,0 +1,114 @@
+- `concurrency` The number of builds that can be running on the host at any one time. Note that this is tracked by the controller rather than the host, so for example if you had a host referenced by both staging and production each environment is not aware of the other environments builds.
+- `user` The username to use to connect to the host
+- `arch` The hosts architecture
+
+
+### Allocation Process
+
+When a `TaskRun` is identified that meets the contract above the controller will attempt to allocate a virtual machine for them.
+
+The controller looks at the current configuration, and also looks for all running tasks with a host assigned. Using this information it picks one of the hosts at random that has the most amount of free slots. This algorithm is very simple, but works fine when all hosts are the same size. If we are mixing hosts of different sizes then we may need a better algorithm.
+
+Once the host is selected the controller runs a 'provisioning' task to prepare the host. This task currently does the following:
+
+- Create a new user specifically for this build
+- Generate a new pair of SSH keys for this user and install them
+- Copy the private key for this user back to the cluster and create the required secret with this information
+
+Once the secret has been created then the original `TaskRun` will start, and can use the new user on the remote host to perform the build. This approach means that every build is run under a different user, and only low privilege ephemeral SSH keys end up in the user's namespace.
+
+Which hosts are allocated with which tasks are tracked by adding additional labels onto the `TaskRun` object.
+
+If the provisioning task fails then a label is added to the `TaskRun` indicating that it has failed and allocation is re-tried, excluding all hosts in the failed list. If all of them fail then the error secret is created.
+
+### Clean Up
+
+When the `TaskRun` is completed then the user is removed from the remote host via a cleanup task.
+
+
+## VM Host Requirements
+
+In order to work with the multi arch controller remote hosts need to meet the following requirements:
+
+- Have podman installed
+- Provide a user that can successfully use `sudo` to add and delete users, and copy generated SSH keys into their home directories
+- They *should* be running RHEL, however this is not strictly speaking a technical requirement
+
+### Additional Requirements
+
+There are still unanswered questions about these hosts and additional requirements that may be placed on them.
+
+- How do we handle logs from these machines? Where will they be stored?
+- Should the 'main' user be locked down so that it can only add and remove the build users? At present the main user used by the controller has root access (not the users used by the build, but the user that creates these users).
+- Do we need any kind of intrusion detection or additional monitoring software on these VMs?
```

Comment:
Yes. Given that these VMs are long-lived, they are a prime target for embedding some kind of solarwind-like malware. @zregvart has pointed me to [this](https://youtu.be/zCznRGlsYvo). We may need something like that.

Reply:
Even as @stuartwdouglas is exploring ephemeral environments, it may still be beneficial to deploy some trusted observer to the environment to help our confidence in the trustworthiness.